### PR TITLE
[WIP] Multiple Service CIDRs

### PR DESCRIPTION
**Pull request by aojea**
/kind feature
/kind api-change

#### What this PR does / why we need it:

- Implements [KEP 1889 - Multiple Service CIDRs](https://github.com/kubernetes/enhancements/tree/master/keps/sig-network/1880-multiple-service-cidrs) , so users can configure the ServiceCIDRs via the API
- Fixes the ServiceCIDR size limitation 2^16 for IPv6. Current limitation is 2^64-1 to avoid dealing with uint64 overflows
- Fixes the bootstrap and "kubernetes.default" Service creation when multiple apiservers use different flags for the service-cidr ranges
- ....

#### Which issue(s) this PR fixes:

Fixes #

#### Special notes for your reviewer:

```
$ kubectl get servicecidr
NAME                       IPV4           IPV6
kubernetes-default-hf5wd   10.96.0.0/16   <none>

$ kubectl get services -A
NAMESPACE     NAME         TYPE        CLUSTER-IP   EXTERNAL-IP   PORT(S)                  AGE
default       kubernetes   ClusterIP   10.96.0.1    <none>        443/TCP                  5m42s
kube-system   kube-dns     ClusterIP   10.96.0.10   <none>        53/UDP,53/TCP,9153/TCP   5m36s

$ kubectl get ipaddresses
NAME         PARENTREF
10.96.0.1    services/default/kubernetes
10.96.0.10   services/kube-system/kube-dns
```

#### Does this PR introduce a user-facing change?

```release-note

```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs
- [KEP]: https://github.com/kubernetes/enhancements/tree/master/keps/sig-network/1880-multiple-service-cidrs

```

### CODE REVIEW

File: api/api-rules/violation_exceptions.list [hunk 1]
```diff
@@ -393,6 +393,9 @@ API rule violation: names_match,k8s.io/api/core/v1,RBDVolumeSource,RBDPool
 API rule violation: names_match,k8s.io/api/core/v1,RBDVolumeSource,RadosUser
 API rule violation: names_match,k8s.io/api/core/v1,VolumeSource,CephFS
 API rule violation: names_match,k8s.io/api/core/v1,VolumeSource,StorageOS
+API rule violation: names_match,k8s.io/api/networking/v1alpha1,ParentReference,Resource
+API rule violation: names_match,k8s.io/api/networking/v1alpha1,ServiceCIDRConfigSpec,IPv4CIDR
+API rule violation: names_match,k8s.io/api/networking/v1alpha1,ServiceCIDRConfigSpec,IPv6CIDR
```
Comment (liggitt) [inline api/api-rules/violation_exceptions.list:398]: we shouldn't add new violations, right?

Comment (aojea) [reply api/api-rules/violation_exceptions.list:398]: I'm just trying to make it work :smile: , there are a lot of things here that are magic to me,

**Comment by aojea** [comment]
ping @thockin for the API review on the first 2 commits

**Comment by aojea** [comment]
lovely, it passes CI e2e :smile:

File: pkg/features/kube_features.go [hunk 2]
```diff
@@ -735,7 +735,14 @@ const (
 	ServiceInternalTrafficPolicy featuregate.Feature = "ServiceInternalTrafficPolicy"
 
 	// owner: @aojea
-	// kep: https://kep.k8s.io/3070
+	// kep: http://kep.k8s.io/1880
+	// alpha: v1.25
+	//
+	// Allow to dynamically configure Service IP ranges
+	ServiceCIDR featuregate.Feature = "ServiceCIDR"
```
Comment (aojea) [inline pkg/features/kube_features.go:742]: s/ServiceCIDR/MultipleServiceCIDR/

File: pkg/features/kube_features.go [hunk 3]
```diff
@@ -775,7 +782,7 @@ const (
 	ServiceInternalTrafficPolicy featuregate.Feature = "ServiceInternalTrafficPolicy"
 
 	// owner: @aojea
-	// kep: https://kep.k8s.io/3070
+	// kep: http://kep.k8s.io/3070
```
Comment (aojea) [inline pkg/features/kube_features.go:785]: https

File: pkg/registry/networking/servicecidr/strategy_test.go [hunk 4]
```diff
@@ -0,0 +1,17 @@
+/*
+Copyright 2021 The Kubernetes Authors.
```
Comment (aojea) [inline pkg/registry/networking/servicecidr/strategy_test.go:2]: 2022

**Comment by aojea** [comment]
back on track,

File: pkg/apis/networking/validation/validation.go [hunk 5]
```diff
@@ -741,3 +743,129 @@ func validateClusterCIDRUpdateSpec(update, old *networking.ClusterCIDRSpec, fldP
...
+// ValidateIPAddressName validates that the name is the decimal representation of an IP address
+func ValidateIPAddressName(name string, prefix bool) []string {
+	var errs []string
+	if prefix {
+		errs = append(errs, "prefix not allowed")
+	}
+	ip, err := netip.ParseAddr(name)
+	if err != nil {
+		errs = append(errs, err.Error())
+	} else if ip.String() != name {
```
Comment (aojea) [inline pkg/apis/networking/validation/validation.go:815]: Ben: we should check the contract if this always return canonical

File: pkg/registry/core/service/ipallocator/service_cidrs.go [hunk 6]
```diff
@@ -0,0 +1,580 @@
...
+	case <-a.startCh:
+	default:
+		return nil, fmt.Errorf("allocator not ready")
+	}
+	ctx := context.Background()
+
+	// allocate an IP from one of the available ServiceCIDRs
+	a.muTree.Lock()
+	cidrs := a.tree.Parents()
+	a.muTree.Unlock()
```
Comment (aojea) [inline pkg/registry/core/service/ipallocator/service_cidrs.go:295]: should we lock the entire operation?

File: pkg/registry/core/service/ipallocator/service_cidrs.go [hunk 7]
```diff
@@ -0,0 +1,580 @@
...
+		// get current allocated IPs associated to this range
+		ipLabelSelector := labels.Set(labelMap).AsSelectorPreValidated()
+		ipAddresses, err := a.ipAddressLister.List(ipLabelSelector)
+		if err != nil {
+			return nil, err
+		}
+		// find an empty address within the range
+		if int64(len(ipAddresses)) >= max {
+			continue
+		}
```
Comment (aojea) [inline pkg/registry/core/service/ipallocator/service_cidrs.go:315]: this was only valid when we used the label to assign ips to cidrs

**Comment by dims** [comment]
This PR has the label work-in-progress, please revisit to see if you still need this, please close if not

**Comment by aojea** [comment]
/close

I'm going to split it in multiple PRs, this is hard to review and/or iterate on

**Pull request closed without merging**